File: usedesk_sdk/src/main/java/ru/usedesk/sdk/external/ui/chat/AttachedFilesAdapter.java
package ru.usedesk.sdk.external.ui.chat;

import android.support.annotation.NonNull;
import android.support.v7.widget.LinearLayoutManager;
import android.support.v7.widget.RecyclerView;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ImageView;

import java.util.ArrayList;
import java.util.List;

import ru.usedesk.sdk.R;
import ru.usedesk.sdk.external.entity.chat.UsedeskFileInfo;
import ru.usedesk.sdk.internal.utils.GlideApp;

class AttachedFilesAdapter extends RecyclerView.Adapter<AttachedFilesAdapter.ViewHolder> {

    private final ChatViewModel chatViewModel;
    private List<UsedeskFileInfo> files = new ArrayList<>();

    AttachedFilesAdapter(@NonNull ChatViewModel chatViewModel,
                         @NonNull RecyclerView recyclerView) {
        this.chatViewModel = chatViewModel;

        recyclerView.setAdapter(this);
        recyclerView.setLayoutManager(new LinearLayoutManager(recyclerView.getContext(), LinearLayoutManager.HORIZONTAL, false));
    }

    void update(@NonNull List<UsedeskFileInfo> attachedFiles) {
        if (this.files != attachedFiles) {
            this.files = attachedFiles;

            notifyDataSetChanged();
        }
    }

    @NonNull
    @Override
    public ViewHolder onCreateViewHolder(@NonNull ViewGroup viewGroup, int i) {
        return new ViewHolder(LayoutInflater.from(viewGroup.getContext())
                .inflate(R.layout.usedesk_item_chat_attached_file, viewGroup, false));
    }

    @Override
    public void onBindViewHolder(@NonNull ViewHolder viewHolder, int i) {
        viewHolder.bind(files.get(i));
    }

    @Override
    public int getItemCount() {
        return files.size();
    }

    class ViewHolder extends RecyclerView.ViewHolder {
        private final ImageView imageViewPreview;
        private final ImageView imageViewDetach;

        private ViewHolder(@NonNull View itemView) {
            super(itemView);

            imageViewPreview = itemView.findViewById(R.id.iv_preview);
            imageViewDetach = itemView.findViewById(R.id.iv_detach);
        }

        private void bind(@NonNull UsedeskFileInfo usedeskFileInfo) {
            if (usedeskFileInfo.getType().equals(UsedeskFileInfo.Type.IMAGE)) {
                GlideApp.with(imageViewPreview)
                        .load(usedeskFileInfo.getUri())
                        .centerCrop()
                        .error(R.drawable.ic_attachment_photo_black)
                        .into(imageViewPreview);
            } else {
                GlideApp.with(imageViewPreview)
                        .load(R.drawable.ic_document_black)
                        .into(imageViewPreview);
            }

            imageViewDetach.setOnClickListener(v -> chatViewModel.detachFile(usedeskFileInfo));
        }
    }
}

File: chat-gui/src/main/java/ru/usedesk/chat_gui/chat/adapters/AttachedFilesAdapter.kt
package ru.usedesk.chat_gui.chat.adapters

import android.view.View
import android.view.ViewGroup
import android.widget.ImageView
import android.widget.TextView
import androidx.lifecycle.LifecycleOwner
import androidx.recyclerview.widget.RecyclerView
import ru.usedesk.chat_gui.R
import ru.usedesk.chat_gui.chat.ChatViewModel
import ru.usedesk.chat_sdk.entity.UsedeskFileInfo
import ru.usedesk.common_gui.*

internal class AttachedFilesAdapter(
        private val chatViewModel: ChatViewModel,
        private val recyclerView: RecyclerView
) : RecyclerView.Adapter<AttachedFilesAdapter.ViewHolder>(), IUsedeskAdapter<ChatViewModel> {

    private var files: List<UsedeskFileInfo> = listOf()

    init {
        recyclerView.adapter = this
    }

    override fun onLiveData(viewModel: ChatViewModel, lifecycleOwner: LifecycleOwner) {
        viewModel.fileInfoListLiveData.observe(lifecycleOwner) {
            onItems(it ?: listOf())
        }
    }

    private fun onItems(attachedFiles: List<UsedeskFileInfo>) {
        if (files != attachedFiles) {
            files = attachedFiles
            notifyDataSetChanged()
            recyclerView.visibility = visibleGone(files.isNotEmpty())
        }
    }

    override fun onCreateViewHolder(viewGroup: ViewGroup, i: Int): ViewHolder {
        return ViewHolder(inflateItem(viewGroup,
                R.layout.usedesk_item_chat_attached_file,
                R.style.Usedesk_Chat_Attached_File) { rootView, defaultStyleId ->
            AttachedFileBinding(rootView, defaultStyleId)
        })
    }

    override fun onBindViewHolder(viewHolder: ViewHolder, i: Int) {
        viewHolder.bind(files[i])
    }

    override fun getItemCount() = files.size

    inner class ViewHolder(
            private val binding: AttachedFileBinding
    ) : RecyclerView.ViewHolder(binding.rootView) {

        fun bind(usedeskFileInfo: UsedeskFileInfo) {
            val previewImageId = binding.styleValues
                    .getStyleValues(R.attr.usedesk_chat_attached_file_preview_image)
                    .getId(R.attr.usedesk_drawable_1)

            showImage(binding.ivPreview,
                    previewImageId,
                    usedeskFileInfo.uri.toString(),
                    ignoreCache = true)

            binding.ivDetach.setOnClickListener {
                chatViewModel.detachFile(usedeskFileInfo)
            }
            binding.tvTitle.text = if (!usedeskFileInfo.isImage()) {
                usedeskFileInfo.name
            } else {
                ""
            }
        }
    }

    internal class AttachedFileBinding(rootView: View, defaultStyleId: Int) : UsedeskBinding(rootView, defaultStyleId) {
        val ivPreview: ImageView = rootView.findViewById(R.id.iv_preview)
        val ivDetach: ImageView = rootView.findViewById(R.id.iv_detach)
        val tvTitle: TextView = rootView.findViewById(R.id.tv_title)
    }
}